### PR TITLE
Support referencing members using %M and MemberName type

### LINKED ISSUE #433: API to reference top-level Kotlin entities, including package name
It would be great to be able to have a reference to a top-level fun or property from another file, and use it like we currently use TypeNames.  I don't know a good name for such a reference - `TopLevelReference` seems excessively vague, whereas `PropertyOrFunReference` is too Springy.  I'm totally open to better ideas!

This hypothetical reference would let us do something like:

```kt
val reference = Reference("com.foo.bar", "doSomething")
val function = FunSpec.builder("somethingElse")
    .returns(String::class)
    .addStatement("return %R()", reference)
    .build()

println(FileSpec.builder("com.example", "Example")
    .addFunction(function)
    .build())
```

and, hopefully, `CodeWriter` would handle the imports just as it does for `%T` placeholders today:

```kt
// Example.kt
package com.example

import com.foo.bar.doSomething

fun somethingElse(): String = doSomething()
```

At first, I thought this might be accomplished by using a `FunSpec` with the name I wanted, but as it happens `FunSpec` doesn't take a package name.  There doesn't seem to be any way in the current API to do this.

### PR DESCRIPTION
**Pull request by Egorand**
Closes #433

### CODE REVIEW

**Comment by Egorand** [comment]
Extra feature to add: resolving name conflicts by importing enclosing types. E.g.:
```kotlin
com.example.HolyGuac.createTaco
com.example.TacoBell.createTaco
```
would currently resolve into:
```kotlin
import com.example.HolyGuac.createTaco

createTaco()
com.example.TacoBell.createTaco()
```
but can be resolved into:
```kotlin
import com.example.HolyGuac.createTaco
import com.example.TacoBell

createTaco()
TacoBell.createTaco()
```

**Comment by Egorand** [comment]
Need reviews!

File: src/main/java/com/squareup/kotlinpoet/MemberName.kt [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+  fun copy(
+    packageName: String = this.packageName,
+    enclosingClassName: ClassName? = this.enclosingClassName,
+    simpleName: String = this.simpleName
+  ): MemberName {
+    return MemberName(packageName, enclosingClassName, simpleName)
+  }
+
+  override fun equals(other: Any?): Boolean {
```
Comment (swankjesse) [inline src/main/java/com/squareup/kotlinpoet/MemberName.kt]: should this be a data class?

File: src/test/java/com/squareup/kotlinpoet/MemberNameTest.kt [hunk 2]
```diff
@@ -0,0 +1,159 @@
+package com.squareup.kotlinpoet
```
Comment (swankjesse) [inline src/test/java/com/squareup/kotlinpoet/MemberNameTest.kt]: nit: copyright 2019

File: src/test/java/com/squareup/kotlinpoet/MemberNameTest.kt [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+      |import com.twitter.tacos.TwitterTacos
+      |
+      |fun makeTastyTacos() {
+      |    createTaco()
+      |    TwitterTacos.Companion.createTaco()
+      |}
+      |""".trimMargin())
+  }
+
+  @Test fun memberNameAliases() {
```
Comment (swankjesse) [inline src/test/java/com/squareup/kotlinpoet/MemberNameTest.kt]: oh hell yes

File: src/test/java/com/squareup/kotlinpoet/MemberNameTest.kt [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+      |import com.twitter.tacos.createTaco as createTwitterTaco
+      |
+      |fun makeTastyTacos() {
+      |    createSquareTaco()
+      |    createTwitterTaco()
+      |}
+      |""".trimMargin())
+  }
+
+  @Test fun keywordsEscaping() {
```
Comment (swankjesse) [inline src/test/java/com/squareup/kotlinpoet/MemberNameTest.kt]: this is a great example

File: src/test/java/com/squareup/kotlinpoet/MemberNameTest.kt [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+      |import org.mockito.`when`
+      |
+      |class TacoTest {
+      |    @Before
+      |    fun setUp() {
+      |        `when`(tacoService.createTaco()).thenReturn(tastyTaco())
+      |    }
+      |}
+      |""".trimMargin())
+  }
```
Comment (swankjesse) [inline src/test/java/com/squareup/kotlinpoet/MemberNameTest.kt]: can you add a test for a collision between a member name and a class name? It’s gross but possible

File: src/main/java/com/squareup/kotlinpoet/MemberName.kt (hunk 1 above)
Comment (Egorand) [reply src/main/java/com/squareup/kotlinpoet/MemberName.kt]: Yep, no reason for it not to

File: src/test/java/com/squareup/kotlinpoet/MemberNameTest.kt (hunk 5 above)
Comment (Egorand) [reply src/test/java/com/squareup/kotlinpoet/MemberNameTest.kt]: Good call!

File: src/main/java/com/squareup/kotlinpoet/CodeWriter.kt [hunk 6]
```diff
@@ -360,6 +367,31 @@ internal class CodeWriter constructor(
...
 
+  fun lookupName(memberName: MemberName): String {
+    val simpleName = memberImports[memberName.canonicalName]?.alias ?: memberName.simpleName
+    // Match an imported member.
+    val importedMember = importedMembers[simpleName]
+    if (importedMember == memberName) {
+      return simpleName
+    } else if (importedMember != null && memberName.enclosingClassName != null) {
+      val enclosingClassName = lookupName(memberName.enclosingClassName)
+      return "$enclosingClassName.$simpleName"
```
[the remark this answers is not in the document]
Comment (Egorand) [reply src/main/java/com/squareup/kotlinpoet/CodeWriter.kt]: Yep, escaping happens inside `MemberName.emit()` after the name has been resolved. Test added.

**Pull request merged**